Negative tof to d-space conversion error

peterfpeterson opened the pull request:
Apparently some instruments still have a DAS/DAQ/DAE that sometimes creates a negative time-of-flight in the raw data file. This demonstrated that the quadratic equation was not correctly converting to d-spacing when `DIFA==0`. This PR adds an extra trap to simplify the calculation in this case. Also changes the order of special cases to move calculating intermediate results until as late as possible. In principle, this should execute the case of `DIFA==0` faster as zero is not squared then square rooted (when using instrument geometry).

This was originally discovered when looking at the results of
```python
from mantid.simpleapi import SNAPReduce

SNAPReduce(RunNumbers='52185', Masking='Horizontal', Binning='0.3,0.003,6',
    Normalization='Extracted from Data', PeakClippingWindowSize=12,
    SmoothingRange=5, SaveData=False)
```
The characteristic of it was that the raw data had a time-of-flight=-1 microseconds which exposed a bug in the time-of-flight to d-space conversion using instrument parameters (i.e. `DIFA=0` and `TZERO=0`). This case is captured in the new checks in the unit test. 

**To test:**

Ideally running the script above with mantid v6.1 and this PR and seeing that there is no longer a difference. In practice, the unit test was broken before change to `Unit.cpp` and is fixed now.

*There is no associated issue.*

*This does not require release notes* because it fixes a defect introduced during the release cycle.

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

peterfpeterson committed:
Fix behavior for negative time-of-flight to d

Apparently some instruments still have a DAS/DAQ/DAE that sometimes
creates a negative time-of-flight in the raw data file. This
demonstrated that the quadratic equation was not correctly converting to
d-spacing when DIFA==0. This PR adds an extra trap to simplify the
calculation in this case. Also changes the order of special cases to
move calculating intermediate results until as late as possible. In
principle, this should execute the case of difa==0 faster as zero is not
squared then square rooted.

KedoKudo commented on Framework/Kernel/src/Unit.cpp:
```diff
@@ -644,56 +644,80 @@ void dSpacing::init() {
       }
     }
   }
+  // force the assumption that difc is positive in debug builds
+  assert(difc > 0.);
```
why not throw a runtime error when encountering negative difc?   It is more informative

peterfpeterson replied:
`assert` only runs in debug builds and the vast majority of times people are going to give this positive values. When something is "wrong" a developer will generally switch to a debug build and this will kick in and tell them that an assumption has been violated without any (normal) runtime performance hit. The other runtime errors are more likely to happen.

KedoKudo commented on Framework/Kernel/src/Unit.cpp:
```diff
@@ -644,56 +644,80 @@ void dSpacing::init() {
     throw std::runtime_error("dSpacingBase::singleToTOF called before object "
                              "has been initialized.");
-  return difa * x * x + difc * x + tzero;
+  if (difa == 0.)
```
does c++ support float point equal comparison?

peterfpeterson replied:
Yes it does. Comparing to most (non-integer) float values normally gets you into the realm of precision. However, this is checking against the value that the constructor sets as the default.

KedoKudo commented on Framework/Kernel/src/Unit.cpp:
```diff
@@ -644,56 +644,80 @@ void dSpacing::init() {
     throw std::runtime_error(toDSpacingError);
-  double c = tzero - tof;
+
+  // force the assumption that difc is positive in debug builds
+  assert(difc > 0.);
```
same as before, how about a runtime_error since a negative difc doesn't make any sense?

DannyHindson replied:
Perhaps you've decided to keep anyway but I think the asserts are not needed because the dSpacing::validateUnitParams method already checks difc is positive?

KedoKudo reviewed (commented):
Looks good to me, just a few non-critical questions.
@DannyHindson Could you also take a look at the changes?  Thanks :pray:

DannyHindson commented on Framework/Kernel/src/Unit.cpp:
```diff
@@ -644,56 +644,80 @@ void dSpacing::init() {
+
+  // don't need to solve a quadratic when difa==0
+  if (difa == 0.)
+    return (tof - tzero) / difc;
```
I did think about whether to do this when I originally wrote the code. I decided not to in order to make the code cleaner and also because the performance hit of going through the full quadratic code seemed minimal. When difa=0, the sqrt is always sqrt(1) and this is quicker than sqrt on other values eg I timed sqrt(1) vs sqrt(M_PI) on MSVC and sqrt(1) was faster. Not sure if it's optimised or the sqrt algorithm just completes quicker on 1.

Even without this when I profiled doing unit conversions on large event workspaces as part of this change, this unit conversion code never showed up anywhere near the top

peterfpeterson replied:
Thank you for the results of the measurement. Have you checked adding in the flops for calculating the stuff that goes into the `std::sqrt` or just sqrt itself?

DannyHindson reviewed (commented):
Thanks for sharing this.

It seems to me that the correct change to handle this scenario (negative TOF values + difa=0) is to modify the validation on L685 of the original code to change the > to a >= on the difa check:
```
  if ((difa >= 0) && (c > 0)) {
    throw std::runtime_error("Cannot convert to d spacing. Quadratic doesn't "
                             "have a positive root");
  }
```
This prevents the user doing a conversion on a negative TOF value in a case where DIFA=TZERO=0 because the resulting d spacing is negative. We raise validation errors to prevent negative d spacing being returned when the full quadratic formula is used so it seems consistent to also do it when difa=0?

To get around this the user would have to introduce a negative tzero term into their calibration. If they don't want to add the complexity of a proper calibration including a tzero term into their reduction workflow I guess they could just set tzero equal to the most negative tof value in the dataset

I've added a couple of other detail level comments but the point above is the main one I think.

I agree the test cases didn't cover this scenario so regardless of the outcome of the discussion on how to fix it we should extend the test

peterfpeterson commented on Framework/Kernel/src/Unit.cpp:
```diff
@@ -644,56 +644,80 @@ void dSpacing::init() {
   }
+
+  // general equation
+  const double linearTerm = tzero - tof;
```
The check for `difa==0` will be exercised whenever people convert from d-spacing using the instrument geometry. This trades a `std::sqrt(0)` for an `if`. Since `difa==0` simplifies to solving a simple linear equation, I didn't see a reason to add in all the extra math. This is analogous to picking the correct root of the quadratic based on the sign of `difa` rather than always solving for both and seeing which one is physical.

"constant" is likely to be misinterpreted, but I see your point. My bigger concern what having a variable named "c" in the neighborhood of a concept called "DIFC" which tripped me up. I'm happy with an name that make me think less. What if we think of the terms as part of an equation of a generalized polynomial `0=a0 + a1 * x + a2 * x^2`?

peterfpeterson replied:
More on your suggetion above for expanding the error message to check for `(difa >=0) && (c > 0)`. The error message would be misleading for the case of `difa==0` since people wouldn't understand why we're solving a quadratic for a "simple" linear conversion.

DannyHindson replied:
I guess we'd need to modify the error message here if we did  `(difa >=0) && (c > 0)`. Perhaps a message saying it's not possible to return a positive d spacing is more user friendly even in the quadratic case

peterfpeterson committed:
Change behavior to generate exception for negative d

This also shuffles the code around a little more to cluster the error
checking and separately cluster the returns

DannyHindson commented on Framework/Kernel/src/Unit.cpp:
```diff
@@ -644,56 +644,87 @@ void dSpacing::init() {
+/**
+ * DIFA * d^2 + DIFC * d + T0 - TOF = 0
+ *
+ * Use the citardauq formula to solve quadratic in order to minimise loss of precision. citarqauq (quadratic spelled
```
It should be citardauq rather than citarqauq here I think - not a pleasant word to spell at the best of times!

DannyHindson reviewed (commented):
As you'll have seen from Martyn's Slack message - he would like to keep the separate return for when difa=0 in singleFromTOF.
I've just had one more look through this and have added a couple of minor comments but hopefully we're nearly done with this and can get this fix into the release. Thanks for spotting\fixing

peterfpeterson committed:
Minor rewording of docs and exception
